sql/lease.go

```diff
@@ -0,0 +1,547 @@
+				s.refcount++
+				return s, nil
+			}
+			if s.expiration-store.clock.Now().WallTime >= minLeaseDuration {
+				s.refcount++
+				return s, nil
+			}
+		} else if version != 0 {
+			n := t.active.findNewest(0)
+			if n != nil && version < n.Version {
+				return nil, util.Errorf("table %d unable to acquire lease on old version: %d < %d",
+					t.id, version, n.Version)
+			}
+		}
+
+		if t.acquiring != nil {
+			// There is already a lease acquisition in progress. Wait for it to complete.
+			t.mu.Unlock()
+			<-t.acquiring
+			t.mu.Lock()
+		} else {
+			// There is no active lease acquisition so we'll go ahead and perform
+			// one.
+			t.acquiring = make(chan struct{})
+			t.mu.Unlock()
+			s, err := store.Acquire(t.id, version)
+			t.mu.Lock()
+			close(t.acquiring)
+			t.acquiring = nil
+			if err != nil {
+				return nil, err
+			}
+			t.active.insert(s)
+		}
+
+		// A new lease was added, so loop and perform the lookup again.
+	}
+}
+
+func (t *tableState) release(lease *LeaseState, store LeaseStore) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	s := t.active.find(lease.Version, lease.expiration)
+	if s == nil {
+		return util.Errorf("table %d version %d not found", lease.ID, lease.Version)
+	}
+	s.refcount--
+	if s.refcount == 0 {
+		n := t.active.findNewest(0)
+		if s != n {
+			if s.Version < n.Version {
+				// TODO(pmattis): If an active transaction is releasing the lease for
+				// an older version, hold on to it for a few seconds in anticipation of
+				// another operation being performed within the transaction. If we
+				// release the lease immediately the transaction will necessarily abort
+				// on the next operation due to not being able to get the lease.
+			}
+			t.active.remove(s)
+			return store.Release(s)
+		}
+	}
+	return nil
+}
+
+// LeaseManager manages acquiring and releasing per-table leases. Exported only
+// for testing.
+type LeaseManager struct {
+	LeaseStore
+	mu     sync.Mutex
+	tables map[ID]*tableState
+}
+
+// NewLeaseManager creates a new LeaseManager.
+func NewLeaseManager(nodeID uint32, db client.DB, clock *hlc.Clock) *LeaseManager {
+	return &LeaseManager{
+		LeaseStore: LeaseStore{
+			db:     db,
+			clock:  clock,
+			nodeID: nodeID,
+		},
+		tables: make(map[ID]*tableState),
+	}
+}
+
+// Acquire acquires a read lease for the specified table ID. If version is
+// non-zero the lease is grabbed for the specified version. Otherwise it is
+// grabbed for the most recent version of the descriptor that the lease manager
+// knows about.
+func (m *LeaseManager) Acquire(tableID ID, version uint32) (*LeaseState, error) {
+	t := m.findTableState(tableID, true)
+	return t.acquire(version, m.LeaseStore)
+}
+
+// Release releases a previously acquired read lease.
+func (m *LeaseManager) Release(lease *LeaseState) error {
+	t := m.findTableState(lease.ID, false)
+	if t == nil {
+		return util.Errorf("table %d not found", lease.ID)
+	}
+	return t.release(lease, m.LeaseStore)
```

Comment:
should `Release` remove the table state if nothing's acquired and it's not acquiring, i.e. should `m.tables` ever be pruned?